### PR TITLE
Store raw installed versions for npm, pnpm, and yarn package managers instead of detected version

### PR DESCRIPTION
**Pull request by kbukum1**
#### What are you trying to accomplish?
This PR updates the package manager version retrieval to store the raw installed version (including pre-release and build metadata, if any) for npm, pnpm, and yarn, instead of defaulting to a selected or guessed version. This change ensures that the exact installed version is captured, reflecting the actual environment more accurately.

#### Anything you want to highlight for special attention from reviewers?
- The `installed_version` method has been modified to retrieve and validate the raw version format directly from the package manager, falling back to the lockfile version if the format is unexpected.
- The memoization now caches the raw version, reducing calls to retrieve the version for the same package manager name.
- This approach improves accuracy by capturing the true installed version, which is essential for debugging and tracking specific package versions in complex environments.

#### How will you know you've accomplished your goal?
- Tests are included to verify that the raw version, including tags or metadata (e.g., `6.4.5-alpha`), is stored and returned when available.
- Performance is maintained through memoization, ensuring minimal repeated calls for version retrieval.
- The functionality is confirmed to work as expected across npm, pnpm, and yarn, with correct fallbacks to lockfile versions if the installed version does not match expected patterns.

---

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Review by abdulapopoola — APPROVED** [review]

File: npm_and_yarn/spec/dependabot/npm_and_yarn/package_manager_helper_spec.rb [hunk 1]
```diff
@@ -118,4 +118,58 @@
...
+    context "when the installed version matches the expected format" do
+      before do
+        allow(Dependabot::SharedHelpers).to receive(:run_shell_command)
+          .with("npm --version", fingerprint: "<name> --version").and_return("7.5.2")
+      end
+
+      it "returns the raw installed version and caches it" do
+        expect(helper.installed_version("npm")).to eq("7.5.2")
+        # Check that the version is memoized
+        expect(helper.instance_variable_get(:@installed_versions)["npm"]).to eq("7.5.2")
```
Comment (landongrindheim) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/package_manager_helper_spec.rb:140]: 🤔 Unless we've identified this as a performance bottleneck, I'm not sure of the value we get from checking this. The benefit I do see, which might not need to be repeated, is that we have a guarantee that `@installed_versions` hasn't been typoed.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb [hunk 2]
```diff
@@ -10,6 +10,7 @@ module NpmAndYarn
     ECOSYSTEM = "npm_and_yarn"
     MANIFEST_FILENAME = "package.json"
     LERNA_JSON_FILENAME = "lerna.json"
+    PACKAGE_MANAGER_VERSION_REGEX = /^\d+\.\d+\.\d+(-[a-zA-Z0-9.]+)?(\+[a-zA-Z0-9.]+)?$/
```
Comment (landongrindheim) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:13]: I'm not great at reading Regex like this. If it seems worth your while, what do you think of naming the captures? I think that would make it 🤏 easier for me to parse. That said, I think this is doing what we need it to 😄

**Review by landongrindheim — APPROVED** [review]

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb (hunk 2 above)
Comment (kbukum1) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:13]: Sorry about that. You mean using something like the following?

```
PACKAGE_MANAGER_VERSION_REGEX = /^(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)(?:-(?<pre_release>[a-zA-Z0-9.]+))?(?:\+(?<build>[a-zA-Z0-9.]+))?$
```

Comment (landongrindheim) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:13]: Yep 😄 _I_ find that much easier to parse. What do you think of it?

Comment (kbukum1) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: Change is made.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb [hunk 3]
```diff
@@ -276,19 +278,44 @@ def setup(name)
...
+          "#{name} --version",
+          fingerprint: "<name> --version"
+        ).strip
+
+        if installed_version.match?(PACKAGE_MANAGER_VERSION_REGEX)
+          @installed_versions[name] = installed_version
+          return @installed_versions[name]
+        end
+
+        @installed_versions[name] = Helpers.send(:"#{name}_version_numeric", @lockfiles[name.to_sym]).to_s
```
[the remark this answers is not in the document]
Comment (kbukum1) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: > 💅 Could we use `public_send`? I don't anticipate there being other methods which would match this pattern higher up the chain, but it seems 🤏 safer.

Thanks for the suggestion.

**Review by landongrindheim — APPROVED** [review]

**Pull request merged**